heap/max_heap.py

```diff
@@ -2,20 +2,77 @@ class Heap:
   def __init__(self):
     self.storage = []
 
+  """adds the input value into the heap; this method should ensure
+  that the inserted value is in the correct spot in the heap
+  """
   def insert(self, value):
-    pass
+    self.storage.append(value)
+    self._bubble_up(len(self.storage) - 1)
 
+  """removes and returns the 'topmost' value from the heap; this method
+  needs to ensure that the heap property is maintained after the topmost
+  element has been removed
+  """
+  # TODO
   def delete(self):
-    pass
+    topmost_value = self.storage.pop(0)
+    self._sift_down(0)
+    self._bubble_up(len(self.storage) - 1)
+    self._bubble_up(len(self.storage) - 2)
+    self._bubble_up(len(self.storage) - 3)
+    self._bubble_up(len(self.storage) - 4)
+    return topmost_value
 
+  """returns the maximum value in the heap in constant time
+  """
   def get_max(self):
-    pass
+    return self.storage[0]
 
+
+  """returns the number of elements stored in the heap
+  """
   def get_size(self):
-    pass
+    return len(self.storage)
 
+  """moves the element at the specified index 'up' the heap by swapping it
+  with its parent if the parent's vaule is less than the value at the spefified index.
+  """
   def _bubble_up(self, index):
-    pass
+    while index > 0:
+      parent = (index - 1) // 2
+      if self.storage[index] > self.storage[parent]:
```

Comment:
Hah, love how simple this is. But does this store the element to the right place if the self.storage[index] is in the right place before that loop happens? or is this method only called to double check if the element is in the right place?